### ISSUE TITLE
Turbo 2 UI: Can't copy terminal outputs

### ISSUE THREAD

**Issue by pmm26** [vercel/turbo#8372]
### Verify canary release

- [X] I verified that the issue exists in the latest Turborepo canary release.

### Link to code that reproduces this issue

turbo dev

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Mac

### Which canary version will you have in your reproduction?

turbo-darwin-arm64@2.0.3-canary.0

### Describe the Bug

After installing the new Turbo Repo V2 there was a moment of exiting in seeying the new UI. Followed shortly by a bit disapointment and reverting back to the old UI.

I have 2 things that would be great to have improved.

**1 - Can't copy/select anything.**
Have you ever copied some long to look on google for a possible fix.
Well I can't copy anything apart from clicking on links that VS Code recognizes.

[**2 - Scrolling is hard and slow.**](https://github.com/vercel/turbo/issues/8643)
Have you ever console.log a whole api response?
Well with this new UI forget it you will never reach the top.

IF you are not at the bottom. The application will continue to print things off screen and if you want to see the latest things. 
It's going to take you a good 5 to 20 seconds to get to what's being run now.

**Suggestions:** 
- Add something to be able to go straight to Latest.
- Scroll bar (not sure if it's possible) or some keybind to scroll up faster 

### Expected Behavior

Be able to scroll quickly and be able to copy text from the terminal.

### To Reproduce

turbo dev

### Additional context

I tested it on canary.

I'm running on a MAC M2 Pro V14.2.1
Terminal running in VS Code.
Running on ZSH: zsh 5.9 (arm-apple-darwin22.6.0)

Running Oh-my-zsh:
ZSH_THEME="robbyrussell"

**Comment by lveillard** [comment]
Up! Its a great V0 but not the best UX yet

My struggles are the same, so duplicating some ideas I support and adding new ones:
- Scrollbar or go to latest. Also enhance scroll performance somehow.
- Being able to copy, at least when interacting after entering with Enter
- Enable terminal ctrl F search
- Resize on window resize

**Comment by lveillard** [comment]
Also had some issues debugging an error that was happening in a different task. 
Used to having all the errors together I totally missed it for hours!

Having some kind of UX that points to tasks that have thrown an error, like coloring the name in red, or having some kind of notification would be great.

Ex: Youre running a nextjs app, and a server, the server throws an error and is not well catched in your nextjs app so you dont see the error as it is in the context of the server task

**Comment by anthonyshew** [comment]
When you say "can't copy", are you meaning the same as https://github.com/vercel/turbo/issues/8306 or are you meaning "can't select at all"?

**Comment by pmm26** [comment]
Yes that is what I mean

> When you say "can't copy", are you meaning the same as #8306 or are you meaning "can't select at all"?

**Comment by pmm26** [comment]
> When you say "can't copy", are you meaning the same as #8306 or are you meaning "can't select at all"?

Kinda, It can't select at all. 
I can't select not even to get dirty output.

**Comment by pmm26** [comment]
> * Scrollbar or go to latest. Also enhance scroll performance somehow.

A scrollbar can solver the problem with going up or down.

a "go to latest" solves the problem of going down. In my opinon it would also need something extra to go up faster.

**Comment by lveillard** [comment]
> > When you say "can't copy", are you meaning the same as #8306 or are you meaning "can't select at all"?
> 
> I can't select at all. Not even to get dirty output.

Same for me. In vscode and terminal is powershell 😅

**Comment by imprakharshukla** [comment]
Same here! Cannot select any text.

**Comment by Crauzer** [comment]
Yep, can't select anything at all from my VS Code terminal. This completely destroys UX because you cannot copy errors etc...

**Comment by jackiboy** [comment]
I reverted back to 1.0 because of the scroll issue. It does not behave like a normal terminal window when scrolling.

**Comment by thhermansen** [comment]
Setting env var `TURBO_UI=false` may help too until the new UI gets improved.

**Comment by iAverages** [comment]
There is a config option in the turbo.json file to toggle back to the old output: https://turbo.build/repo/docs/reference/configuration#ui

**Comment by leonard-henriquez** [comment]
Although this new UI is a great idea, it seems half baked.
Currently, if we can't scroll, select and copy data from the output, it's a productivity killer.
Going back to the old UI but can't wait to have a fully working version !
(it's awesome that turbo is open source but it's a real bummer that it doesn't rely more on the community to get feedback before launching and releasing new features)

**Comment by anthonyshew** [comment]
@leonard-henriquez This feature was the first to go through the official RFC process: https://github.com/vercel/turbo/discussions/7802

Bugs are bugs and we are working hard at getting fixes.

**Comment by leonard-henriquez** [comment]
I truly appreciate all the open-source work that has been put into Turbo.
It’s great that you implemented an RFC process! 💯
But wasn’t scrolling literally the very first feedback?
I just think it would have been worth waiting for a more stable version before officially releasing v2 and encouraging people to move to the new version.
Anyways, thank you very much for working on it!

**Comment by mxn2020** [comment]
The console ui is not very useful. It uses a lot of ressources, hangs, does not respond,...
When you launch multiple tasks the resource usage is out of control.

![Screenshot 2024-06-22 at 19 15 57](https://github.com/vercel/turbo/assets/128184210/b47f3150-cbf7-4e2c-9cf8-55e804645f51)

**Comment by chris-olszewski** [comment]
@mxn2020 Can you open an issue with a reproduction. This seems like a very different issue from the original one reported

**Comment by lveillard** [comment]
> The console ui is not very useful. It uses a lot of resources, hangs, does not respond,... When you launch multiple tasks the resource usage is out of control.

Same for me since i've updated turbo to v2 it consumes way more resources and my poor laptop gets too hot. Please add me in the issue you create.

Meanwhile, coming back to the main thread, is there an ETA for being able to copy the console contents? Is in the roadmap? This is the main issue we face with this new version and makes debugging really hard in several scenarios. And indeed the scroll speed is the second most painful regression :S

**Comment by lveillard** [comment]
Btw I know its late for this kind of feedback, but,  why not just running N terminal sessions instead of a single one with this complicated feature? Just opening N terminals in the right order would do.  Since I've updated to turbo 2 thats what Im actually doing while these features are not there. I go one by one to the specific folders and run the dev commands. This way I have the 3-4 tasks running

![image](https://github.com/vercel/turbo/assets/29461076/3848fecb-1be6-456e-a331-2c007916156f)

I feel like building this full single multi-terminal from scratch could be a bit overkill righr?

**Comment by anthonyshew** [comment]
Hey, folks, updating here:

We've disabled the terminal UI in `2.0.6` as the default until we can sort out this issue specifically. We're estimating around a week to get it fixed, and it's clear this isn't the experience we were looking for.

We're still focused on getting the fix for this and appreciate the kind feedback on this. We'll be back.

As a reminder, options to toggle the TUI on and off at your convenience:
- The [`TURBO_UI=1` and `TURBO_UI=0`](https://turbo.build/repo/docs/reference/system-environment-variables), respectively
- [Configuration in `turbo.json`](https://turbo.build/repo/docs/reference/configuration#ui) as `"ui": "tui" | "stream"`

**Comment by samburgers** [comment]
for the scrolling, this terminal ui is pretty decent for scrollbar handling inspiration


https://github.com/jesseduffield/lazygit

**Comment by anthonyshew** [comment]
In the interest of ensuring we're keeping issues focused on one thing at a time, I've created an issue to focus on scrolling here: https://github.com/vercel/turbo/issues/8643

Those of us on the core team will be focusing on copying terminal outputs on this issue. For feedback on scrolling, please visit #8643.

We appreciate those of you contributing feedback on this Issue and appreciate you keeping up with our housekeeping. 🙏

**Comment by lishaduck** [comment]
This isn't entirely the same, but it's the same root cause—the border, which also breaks links. When in, i.e., vscode, you can ⌘-click links from eslint, etc, but this wraps them and breaks longer links.

**Comment by anthonyshew** [comment]
@lishaduck You're correct that that wouldn't be in scope for this issue. If you'd like to report it in a new one, that would be great.

**Comment by anthonyshew** [comment]
Fixed in #8713 and released in [2.0.10](https://github.com/vercel/turbo/releases/tag/v2.0.10)! We appreciate everyone's patience on this as we got this figured out.

**Comment by duard** [comment]
how ca we disable this ?

**Comment by anthonyshew** [comment]
@duard, you can control your UI preference using [the `ui` config](https://turbo.build/repo/docs/reference/configuration#ui) or the `--ui` flag (sorry for lack of link, looks like my redeploy of the docs didn't go through on Friday). 😄 

@shawnmclean, we're tracking improving scrolling [here](https://github.com/vercel/turbo/issues/8643) if you'd like better visibility.

**Comment by net** [comment]
I agree with @lveillard, reimplementing tmux is the wrong direction. The DX will always be subpar compared to what the user is used to from their actual terminal.

Instead the user should just open N terminal sessions so they can get the full features of whatever environment they run their terminal sessions in (be it VSCode, tmux, Zellij, Zed, just their native terminal app, etc.) and the Turborepo daemon should deduplicate task runs across the multiple sessions behind the scenes.

**Comment by gitchiefdeals** [comment]
resurrected in 2.4.2

**Comment by steezeburger** [comment]
> resurrected in 2.4.2

Yep, experiencing this issue again on 2.4.2 as well.

**Comment by anthonyshew** [comment]
@gitchiefdeals @steezeburger, I'm not able to trivially reproduce, and I'd be surprised if this were the case because we haven't touched these code paths recently. If you're able, please create a new Issue with a reproduction.

**Comment by RawToast** [comment]
If it helps, I noticed this in the past but it seems fixed in `2.5.4`

**Comment by NayamAmarshe** [comment]
Facing the same issue. Unable to copy anything from the terminal in v2.5.4 on VSCode.

**Comment by anthonyshew** [comment]
@NayamAmarshe I am unable to reproduce.

https://github.com/user-attachments/assets/77e7307e-d45e-46f8-9a42-01482df857a9

**Comment by arnoldsandoval** [comment]
@anthonyshew I can select it, but i cant actually copy it

**Comment by anthonyshew** [comment]
@arnoldsandoval Pressing `c` on your keyboard doesn't copy the text? I'm not able to reproduce this.

**Comment by nonlinearthink** [comment]
> [@NayamAmarshe](https://github.com/NayamAmarshe) I am unable to reproduce.
> 
>  CleanShot.2025-06-21.at.05.13.39.mp4

I was still fine a few days ago, but two days ago I found out that the content I copied was not the content I selected, I can't tolerate this today, finally switching from `tui` to `stream` solved this problem. My environment: `macOS 15.5`, `apple silicon`, `pnpm`, use `pnpm dlx create-turbo@latest` on my computer can reproduce this issue, I tried terminal app, iTerm2, VSCode, cursor, and switched to zsh or bash, it's all the same.

**Comment by Sporarum** [comment]
Here is what I have, on `macOS 14.6`, `Apple M3`, `pnpm`, in the vscode terminal:

https://github.com/user-attachments/assets/8ed59a18-fc11-46c7-b63f-43bb7a554e33

Neither command + c, nor shift + command + c work

**Comment by aifirstd3v** [comment]
Cannot copy with iterm2. I set everything right with iterm2 to copy stuff. I cannot copy anything only with turborepo.

**Comment by President-Q** [comment]
> Here is what I have, on `macOS 14.6`, `Apple M3`, `pnpm`, in the vscode terminal:
> 
>  Screen.Recording.2025-07-24.at.10.29.49.mov 
> Neither command + c, nor shift + command + c work

I believe it's just pressing 'c'. No other combinations.

**Comment by Sporarum** [comment]
I had in the meantime discovered that, it is however very un-intuitive that in the terminal I can copy everywhere with command + `c`, except for turbo

**Comment by pixelmultiplo** [comment]
What a terribile mess.

**Comment by rafiksh** [comment]
Hey, clicking `c` alone works as @President-Q said :) [macOS 15.6.1, Apple M1]

**Comment by lveillard** [comment]
I feel like making turborepo open N different terminals would have been way easier and built in a single day 😅

**Comment by net** [comment]
I never use the TUI. Works much better with `"ui": "stream"`.

**Comment by lydamien** [comment]
It seems indeed that neither `ctrl+c` nor `ctrl+shift+c` work, but simply `c` does.
If you can't select the terminal content with `"ui":"tui"`, try changing the terminal in VSCode (bash, powershell, etc).

**Comment by steveninety** [comment]
If only i knew that "c" works... I've literally been using screenshots and pasting then into cursor. What a mess. Plus, I'm using claude on wsl so I can't even paste images into the terminal... Very relieved to finally got this solved.

**Comment by cereallarceny** [comment]
@anthonyshew Is there any possibility to re-open this issue? It seems there's a fairly reasonable consensus that just using `c` isn't really normal. While copying definitely works with the `c` character, there's not much precedent for this. Perhaps another issue could be opened that proposes a more normal copy command?

**Comment by adbutterfield** [comment]
I was just going to open a new one, but don't have time to create a reproduction required for a new issue right now...

My problem is I'm also running a test watcher, and you can interact with that terminal to do things like filtering which tests to run by typing the name of the test file. That is, unless there is a "c" in the name of the test file...

**Comment by DominikScholz** [comment]
> [@anthonyshew](https://github.com/anthonyshew) Is there any possibility to re-open this issue? It seems there's a fairly reasonable consensus that just using `c` isn't really normal. While copying definitely works with the `c` character, there's not much precedent for this. Perhaps another issue could be opened that proposes a more normal copy command?

so much this, there ideally both should just be possibel

**Comment by holdenmatt** [comment]
I was hoping this would get fixed (remembering to use "c" in this one context and Cmd+c everywhere else is too much cognitive friction). Reverted to "ui": "stream" until then.

**Comment by lveillard** [comment]
> I was hoping this would get fixed (remembering to use "c" in this one context and Cmd+c everywhere else is too much cognitive friction). Reverted to "ui": "stream" until then.

i really don't understand why building a full terminal from scratch, they could have just a for loop and open N terminals automatically

**Comment by frankdilo** [comment]
still a problem on macOS with Ghostyy

**Comment by kaminskypavel** [comment]
hope this gets fixed in 2026 👋

**Comment by anthonyshew** [comment]
Folks, just want to let everyone know that there are two substantive changes to the TUI experience that affect what is being discussed here.

TL;DR: Highlighting text in the TUI will automatically copy to your clipboard, no `c` press necessary. You can also toggle between the TUI and stream modes by pressing `s` and `h`, so you can use `Ctrl+c` to heart's content in stream mode.

### You can now easily toggle from TUI to stream (and back again) with `s` and `h`

https://github.com/vercel/turborepo/pull/13203

If you are using the terminal UI, you now have the ability to get back to your raw terminal in two ways:
1. Pressing `s` (as in `s`witch), will toggle you back and forth between the TUI and log streaming. This puts *logs for all tasks* back into the raw terminal buffer, so you can use your native terminal features.
2. Pressing `h` (as in `h`ide) will toggle the sidebar. This puts *logs for the selected task* into the raw terminal buffer. Again, you can use your native terminal features.

What do I mean by native terminal features?
- Uses the terminal's native scrollback
- `Ctrl+c` copies text again
- `Ctrl+f` search through not just what is present in the current window, but through scrollback as well
- Anything else that your terminal does in raw mode (For those who don't know much about terminals "raw mode" means "the way you're used to your terminal working".)

This will feel much more natural, and continues the TUI's improvements over `"ui": "stream"`.

### The TUI now automatically copies to clipboard when you release selection

https://github.com/vercel/turborepo/pull/13208

Our terminal UI now uses a common technique for terminal tooling that blocks `Ctrl+C` (because they render to alt screen). We automatically copy the selected text to your clipboard when you release the mouse click. You no longer have to press `c` to get things copied.

If you want to prevent the automatic copying, you can hold `Shift` and _then_ release, and the text will stay highlighted, and the copy will not happen.
